### ISSUE TITLE
Remove support for `extern "rust-intrinsic"` blocks

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#132735]
We currently have two ways to declare symbols that are invoked as intrinsics. The old way:
```rust
extern "rust-intrinsic" {
    fn unreachable() -> !;
}
```
The new way:
```rust
#[rustc_intrinsic]
unsafe fn unreachable() -> !;
```

The goal of this issue is to remove support for the old style, and consistently use the new style.
- [x] Port the remaining `extern "rust-intrinsic"` intrinsics in `library` to the new style, updating them using the pattern described above. This can be a PR on its own. (Implemented in https://github.com/rust-lang/rust/pull/132907, https://github.com/rust-lang/rust/pull/133106, https://github.com/rust-lang/rust/pull/136005)
- [x] Port the uses of `extern "rust-intrinsic"` in `src/tools/miri` and `tests/ui/simd` to the new style. 
   This is implemented in https://github.com/rust-lang/rust/pull/135333, https://github.com/rust-lang/rust/pull/136022.
- [x] Similarly port the `extern "rust-intrinsic"` in [this file](https://github.com/rust-lang/stdarch/blob/master/crates/core_arch/src/powerpc/altivec.rs) (note that this is a separate repository)
- [x] Remove support for `extern "rust-intrinsic"` blocks from the compiler -- in particular, remove [this](https://github.com/rust-lang/rust/blob/eca17022ef267f9ed87ba7d22755a1bfe0082e80/compiler/rustc_abi/src/extern_abi/mod.rs#L203-L206). A bunch of tests will still need updating; you can grep for `rust-intrinsic` to find them all. They should all be ported to the new style, similar to the PR in step 2.
- [ ] Determine whether there is more cleanup that can be done. For instance, [this](https://github.com/rust-lang/rust/blob/6ba39f7dc709a90bcb125974964a7e464ed86ff2/compiler/rustc_hir_analysis/src/check/check.rs#L835-L845) can be removed. Maybe we can get rid of `Abi::RustIntrinsic` entirely?

**Comment by compiler-errors** [comment]
I do wonder if there's some way we could have body-less free function items when users write `#[rustc_intrinsic_must_be_overridden]`. Having all these dummy bodies like `loop {}` and `unreachable {}` seem kinda meh.

**Comment by RalfJung** [comment]
Yeah that would be nice. But I guess the parser will reject body-less fn items (with ";" instead of a block) pretty early?

**Comment by compiler-errors** [comment]
We actually reject them after expansion but before ast lowering. 

We could perhaps intercept these bodyless functions during ast lowering and fill them with a dummy body like `loop {}` or something that diverges rather than having it be manually written on each function.

**Comment by compiler-errors** [comment]
Obviously only intercepting them if they're marked with `rustc_intrinsic_must_be_overridden`; if it's a regular function we'd issue the regular error for a bodyless free function, and if it's `rustc_intrinsic` we could tell the user that it needs to be marked with `rustc_intrinsic_must_be_overridden`.

**Comment by RalfJung** [comment]
That's way beyond my knowledge of those parts of the compiler. ;) So sure, sounds great. :D

Cc @petrochenkov 

> Obviously only intercepting them if they're marked with rustc_intrinsic_must_be_overridden

I was actually doing to suggest that a `#[rustc_intrinsic]` not having a body should replace the `rustc_intrinsic_must_be_overridden` attribute.

**Comment by compiler-errors** [comment]
Oh, that works too.

**Comment by workingjubilee** [comment]
hmm do we have the notion of a builtin attribute macro...?

**Comment by compiler-errors** [comment]
@workingjubilee: Yes

**Comment by BLANKatGITHUB** [comment]
Ok do basically I tried porting the first block of `extern "rust-intrinsic" ` in `library/core/src/intrinsics/mod.rs`
and it gave the errors 

`error[E0308]: intrinsic has wrong type
    core/src/intrinsics/mod.rs:1277:26
    
    pub fn prefetch_read_data<T>(data: *const T, locality: i32) {
                                                    ^^^ expected unsafe fn, found safe fn
     
     note: expected signature `unsafe fn(_, _)`
                found signature `fn(_, _)` 

and 

``
error: intrinsic safety mismatch between list of intrinsics within the compiler and core library intrinsics for intrinsic prefetch_write_data
    core/src/intrinsics/mod.rs:1292:1
      
pub fn prefetch_write_data<T>(data: *const T, locality: i32) {

``

**Comment by RalfJung** [comment]
Yes, the intrinsics need to be marked as `unsafe fn`. Sorry I forgot that when writing up the issue.

**Comment by RalfJung** [comment]
> I do wonder if there's some way we could have body-less free function items when users write `#[rustc_intrinsic_must_be_overridden]`. Having all these dummy bodies like `loop {}` and `unreachable {}` seem kinda meh.

https://github.com/rust-lang/rust/pull/135031 implements that.

**Comment by RalfJung** [comment]
The current status for this issue is that many but not all of the intrinsics in `library` have been ported to the new style.

@BLANKatGITHUB do you plan to continue this, porting the remaining intrinsics in `library`?

**Comment by BLANKatGITHUB** [comment]
Sorry I am currently busy with my exams, but I will get to it as soon I get the time

**Comment by RalfJung** [comment]
No worries. :) Focus on your exams first!

**Comment by RalfJung** [comment]
Note that uses *outside* `library` can use an even newer style now, but it will take 6 weeks before we can use that style in `library/`:
```rust
#[rustc_intrinsic]
unsafe fn unreachable();
```

**Comment by BLANKatGITHUB** [comment]
Thanks for understanding I will be free again around 20 January, I will contact you then for more information.

**Comment by vayunbiyani** [comment]
> Thanks for understanding I will be free again around 20 January, I will contact you then for more information.

Hi @BLANKatGITHUB , I hope your exams went well! While you were away, I made some progress on the issue. Let me know if there's anything else I can assist you with regarding its resolution.

**Comment by BLANKatGITHUB** [comment]
@vayunbiyani thanks I saw you change quite a few files in one pr , I am not quite sure if every file in compiler and all the tests   are ported , I hope you can provide more info on that as I was busy with exams currently I am working on ``` /home/blanky/projects/opensource/rust/library/core/src/intrinsics/mod.rs ``` .

**Comment by BLANKatGITHUB** [comment]
@RalfJung I don't see  ```./library/core/src/ffi/va_list.rs ``` are they ported? and can you provide more info on new style we can't  use it inside the library though, right?

**Comment by vayunbiyani** [comment]
>I am not quite sure if every file in compiler and all the tests are ported

@BLANKatGITHUB You’re correct, not every file has been updated. I’ve only ported some of the more obvious ones. I’m a bit uncertain about how to handle porting the test cases while adhering to best practices. What do you think about jumping on Zulip to figure this out?

**Comment by RalfJung** [comment]
`tests/ui/simd` has a bunch of test cases that need porting, I would suggest those be ported in a separate PR before we do the final big PR that removes support for this ABI from the compiler entirely.

**Comment by BLANKatGITHUB** [comment]
@vayunbiyani sure drop the link

**Comment by vayunbiyani** [comment]
> [@vayunbiyani](https://github.com/vayunbiyani) sure drop the link

You can find me [here](https://rust-lang.zulipchat.com) as `Vayun Biyani`

**Comment by RalfJung** [comment]
Okay I think all the preparation is done now. :) There are fairly few such intrinsic blocks still left in the test suite; this is probably best done together with the actual removal.

**Comment by RalfJung** [comment]
@BLANKatGITHUB do you still plan to work on the next step for this issue, or should the issue assignment be released?

**Comment by BLANKatGITHUB** [comment]
@RalfJung  sorry I was a little busy with college , I just used grep on rust repo and found several files which uses old declaration , I am not sure where to start many are in tests and some in compiler

**Comment by RalfJung** [comment]
As I said above, the next step is to actually remove the "rust-intrinsic" ABI string from the compiler. The remaining uses of the old style can be fixed as part of that PR, there's not so many of them. The PR description even points to the line of code you have to change for that.

**Comment by RalfJung** [comment]
Ah, never mind, there's one bit of cleanup that has to happen before -- there's two uses in `library/stdarch/crates/core_arch/src/powerpc/altivec.rs` that need to be updated via a separate PR in https://github.com/rust-lang/stdarch/.

**Comment by BLANKatGITHUB** [comment]
hey @RalfJung I was wondering as mentioned in issue the removal of [this](https://github.com/rust-lang/rust/blob/eca17022ef267f9ed87ba7d22755a1bfe0082e80/compiler/rustc_abi/src/extern_abi/mod.rs#L203-L206) well I couldn't find this file on main branch am I supposed to change it on master branch or a particular one?

**Comment by RalfJung** [comment]
Seems like the file got renamed, grep for `enum ExternAbi` to find the current location.

Also, https://github.com/rust-lang/stdarch/pull/1722 has to be synced to the rustc repo before you can proceed. I am hoping to include another PR in that sync before I will try a sync myself, though someone else might also do a sync in the mean time.

**Comment by RalfJung** [comment]
@BLANKatGITHUB the stdarch update landed, so this is ready for the next step :)

**Comment by RalfJung** [comment]
@Skgland you opened https://github.com/rust-lang/rust/pull/139455. Given that the issue here still has someone assigned, please always ask the assignee (@BLANKatGITHUB) in the issue first before assuming they have stopped working on this!

**Comment by BLANKatGITHUB** [comment]
Well , I was busy with college, that's quite a big pr though, does this pr finally close the issue? Or there are some more files left

**Comment by RalfJung** [comment]
No worries. I hope you're fine with someone else taking over. :)

That PR entirely removes "rust-intrinsics" from the compiler so it has to fix all the remaining files. The only thing that's left is the last (5th) checkbox in the issue description.

**Comment by Skgland** [comment]
Besides looking for further cleanup potential things still todo would include
- updating [the nomicon Foreign Calling Conventions list](https://doc.rust-lang.org/nomicon/ffi.html#foreign-calling-conventions) in the https://github.com/rust-lang/nomicon repo.
- update the docs at https://github.com/rust-lang/rust/blob/7dd57f085cab4cc671b01d6dd27977c76d04de28/library/core/src/intrinsics/mod.rs#L11 were I left a fixme as I didn't know hot to reword this


And I noticed is that rust-analyser still checks for `#[rustc_must_be_overridden]` though this has been [removed from the compiler](https://github.com/rust-lang/rust/commit/6eea027aa9453b685ac636ea40fc826267afab2a) as its been superseded by intrinsics functions without bodies.
So the check e.g. at https://github.com/rust-lang/rust/blob/7dd57f085cab4cc671b01d6dd27977c76d04de28/src/tools/rust-analyzer/crates/hir-ty/src/mir/eval/shim.rs#L73 and symbol defined at https://github.com/rust-lang/rust/blob/7dd57f085cab4cc671b01d6dd27977c76d04de28/src/tools/rust-analyzer/crates/intern/src/symbol/symbols.rs#L445 can also probably be removed.

**Comment by Skgland** [comment]
> Regarding the nomicon, could you make a PR?

Opened rust-lang/nomicon#485

**Comment by RalfJung** [comment]
@oli-obk do you know of anything left to do here? I think it is all done, isn't it?

**Comment by oli-obk** [comment]
Nope, this is done
